File: LeetCode/rotate_image.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int size = matrix.size();
        for(int i=0; i < size/2; i++){
            for(int j=0; j < size/2; j++){
                rotate_cycle(matrix, i, j);
            }
        }
        if(size%2){
            for(int i=0; i < size/2; i++){
                rotate_cycle(matrix, i, size/2);
            }
        }
        output(matrix);
    }
private:
    void rotate_cycle(vector<vector<int>>& matrix, int i, int j){
        int size = matrix.size();
        int tmp = matrix[i][j];
        matrix[i][j] = matrix[size-1-j][i];
        matrix[size-1-j][i] = matrix[size-1-i][size-1-j];
        matrix[size-1-i][size-1-j] = matrix[j][size-1-i];
        matrix[j][size-1-i] = tmp;
    }
    void output(vector<vector<int>>& matrix){
        int size = matrix.size();
        for(int i=0; i<size; i++){
            for(int j=0; j<size; j++){
                cout << matrix[i][j] << " ";
            }
            cout << endl;
        }
    }
};

int main(void){
    vector<vector<int>> matrix = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };
    Solution sol;
    sol.rotate(matrix);
    return 0;
}

File: LeetCode/rotate_image.py
class Solution(object):
    def rotate(self, matrix):
        size = len(matrix) - 1
        for i in range((size + 1) // 2):
            for j in range(i, size - i):
                tmp = matrix[j][i]
                matrix[j][i] = matrix[size - i][j]
                matrix[size - i][j] = matrix[size - j][size - i]
                matrix[size - j][size - i] = matrix[i][size - j]
                matrix[i][size - j] = tmp

if __name__ == '__main__':
    matrix = [[1,2,3],[4,5,6],[7,8,9]]
    sol = Solution()
    sol.rotate(matrix)
    for line in matrix:
        print(line)